benchmarks/src/bin/wal.rs

```diff
@@ -0,0 +1,394 @@
+            WalOptions::Kafka(KafkaWalOptions {
+                topic: topics.get(id as usize % topics.len()).cloned().unwrap(),
+            })
+        }
+        WalProvider::RaftEngine => WalOptions::RaftEngine,
+    };
+    Region::new(
+        RegionId::from_u64(id),
+        build_schema(&parse_col_types(&cfg.col_types), rng),
+        wal_options,
+        cfg.num_rows,
+        cfg.rng_seed,
+    )
+}
+
+fn build_schema(col_types: &[ColumnDataType], mut rng: &mut SmallRng) -> Vec<ColumnSchema> {
+    col_types
+        .iter()
+        .map(|col_type| ColumnSchema {
+            column_name: Alphanumeric.sample_string(&mut rng, 5),
+            datatype: *col_type as i32,
+            semantic_type: SemanticType::Field as i32,
+            datatype_extension: None,
+        })
+        .chain(vec![ColumnSchema {
+            column_name: "ts".to_string(),
+            datatype: ColumnDataType::TimestampMillisecond as i32,
+            semantic_type: SemanticType::Tag as i32,
+            datatype_extension: None,
+        }])
+        .collect()
+}
+
+fn dump_report(cfg: &Config, write_elapsed: u128, read_elapsed: u128) {
+    let cost_report = format!(
+        "write costs: {} ms, read costs: {} ms, append_batch costs: {} ms, produce costs: {} ms",
+        write_elapsed,
+        read_elapsed,
+        APPEND_BATCH_ELAPSED_TOTAL.load(Ordering::Relaxed),
+        PRODUCED_ELAPSED_TOTAL.load(Ordering::Relaxed),
+    );
+
+    let total_written_bytes = metrics::METRIC_WAL_WRITE_BYTES_TOTAL.get();
+    let write_throughput = if write_elapsed > 0 {
+        total_written_bytes as f64 / write_elapsed as f64 * 1000.0
+    } else {
+        0.0
+    };
+    // This is the effective read throughput from which the read amplification is removed.
+    let total_read_bytes = metrics::METRIC_WAL_READ_BYTES_TOTAL.get();
+    let read_throughput = if read_elapsed > 0 {
+        total_read_bytes as f64 / read_elapsed as f64 * 1000.0
+    } else {
+        0.0
+    };
+
+    let throughput_report = format!(
+                "total written bytes: {} bytes, total read bytes: {} bytes, write throuput: {} bytes/s ({} mb/s), read throughput: {} bytes/s ({} mb/s)",
+                total_written_bytes,
+                total_read_bytes,
+                write_throughput.floor() as u128,
+                (write_throughput / (1 << 20) as f64).floor() as u128,
+                read_throughput.floor() as u128,
+                (read_throughput / (1 << 20) as f64).floor() as u128,
+            );
+
+    let metrics_report = if cfg.report_metrics {
+        let mut buffer = Vec::new();
+        let encoder = TextEncoder::new();
+        let metrics = prometheus::gather();
+        encoder.encode(&metrics, &mut buffer).unwrap();
+        String::from_utf8(buffer).unwrap()
+    } else {
+        String::new()
+    };
+
+    info!("Benchmark config: {:?}\n\nBenchmark report:\n\n{cost_report}\n\n{throughput_report}\n\n{metrics_report}", cfg);
+}
```

Comment:
Maybe this is more readable...
```suggestion
    let result = format!(r#"
Benchmark config: {cfg:?}
Benchmark report:
{cost_report}

{throughput_report}

{metrics_report}
"#);
    info!("{}", result);
```